localstack/plugin/core.py

```diff
@@ -0,0 +1,133 @@
+import abc
+import inspect
+from typing import Any, Callable, Dict, List, Tuple, Type, Union
+
+
+class Plugin(abc.ABC):
+    """A generic LocalStack plugin.
+
+    A Plugin's primary function is to make it easy to be discovered, and to defer code imports into the Plugin::load
+    method. Abstract subtypes of plugins (e.g., a LocalstackCliPlugin) may hook into
+
+    Internally a Plugin is a wrapper around a setuptools EntryPoint. An entrypoint is a tuple: name, module:object
+    inside a namespace that can be loaded. The entrypoint object of a LocalStack Plugin should always point to
+    Plugin.__init__ (the constructor of the Plugin). Meaning that, loading the entry point is equivalent to
+    instantiating the Plugin. A PluginLoader will then run the Plugin::load method.
+    """
+
+    namespace: str
+    name: str
+    requirements: List[str]
+
+    def is_active(self) -> bool:
+        # FIXME: remove after release (would currently break localstack-ext)
+        return self.should_load()
+
+    def should_load(self) -> bool:
+        return True
+
+    def load(self, *args, **kwargs):
+        """
+        Called by a PluginLoader when it loads the Plugin.
+        """
+        return None
+
+
+PluginType = Type[Plugin]
+PluginFactory = Callable[[], Plugin]
+
+
+class PluginSpec:
+    namespace: str
+    name: str
+    factory: PluginFactory
+    metadata: Dict[str, Any]
+    requirements: List[str]
+
+    def __init__(
+        self,
+        namespace: str,
+        name: str,
+        factory: PluginFactory,
+        metadata: Dict[str, Any] = None,
+        requirements: List[str] = None,
+    ) -> None:
+        super().__init__()
+        self.namespace = namespace
+        self.name = name
+        self.factory = factory
+        self.requirements = requirements or []
+        self.metadata = metadata or {}
+
+    def __str__(self):
+        return "PluginSpec(%s.%s = %s)" % (self.namespace, self.name, self.factory)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __eq__(self, other):
+        attrs = [
+            "namespace",
+            "name",
+            "factory",
+            "metadata",
+            "requirements",
+        ]
+
+        for attr in attrs:
+            if getattr(self, attr) != getattr(other, attr):
```

Comment:
i removed all the dynamic code stuff and replaced it with static checks. i think that is better anyway since that makes it more explicit and there shouldn't be confusion about which attributes exist.